### PR TITLE
[Sapling] Anchors and nullifiers integrated into the chain state view layered cache.

### PR DESCRIPTION
**Pull request by furszy**
One more PR decoupled from #1798 to make the review process faster.

Essentially, the chain view state will be storing and caching the best anchor hash and 2 new entry types: (1) Anchors, (2) Nullifiers (Each of them with their entry class and map).

Note: This PR is introducing the new capabilities and validating the functionality correct behavior with a good amount of unit tests (test coverage inside `coins_tests.cpp`), it's not connected to the network flow.

Note2: There is a functional tests in 1798 (`sapling_wallet_anchorfork.py`) that wasn't able to decouple here due larger features requirements (new RPC calls introduced down 1798 commits line) but can be easily run there.

#### List of commits gathered here from #1798:

* CoinsView, Anchor and nullifier view integration. —> 5a8f5ebd3b2d111572aa48ad34717fdb3e84bcbf
* CCoinsViewCache add missing SetNullifiers method —> 1e3aaf191fe4c0a64322239e1e704f475c7ab840
* coins_tests unit test back ported, up-to-date —> 3d46b5c015c0adf459df0a6a0a216012cbbcfba7
* Connect tx GetShieldedValueIn —> 1050850cb9baf3a055a7337f6c39050f95e98aaa [PARTIALLY, only the coins view side. No wallet changes]
* Missing optional sapData correction —> 8afcdd676058b40265fb56121970ee4f5661baee
* Add sapling tx priority to the GetPriority calculation —> 6ba51f2d002f8a62615645a90f81a9f948cb16a8
* fees: Do estimate fee/priority for shielded transaction. Those have fixed fee and priority. —> 84d61dfe24be82ce62195543a6efb62c7c551c5f
* coins_tests: remove not needed BOOST_TEST_CONTEXT. We only have Sapling tests. —> 94aca901b53f1ebae755bb3571a6ea12e41adc7e

### CODE REVIEW

**Comment by furszy** [comment]
Rebased on top of master, conflicts solved. Ready for review

**Review by random-zebra — COMMENTED** [review]
Pretty good. Code ACK with few notes.

Just an idea for the future:
In the tests here we usually do:
```cpp
    CCoinsViewTest base;
    CCoinsViewCache cache1(&base);
```
and then operate on the `CCoinsViewCache` object `cache1`.
It would be good, instead, to layer `CCoinsViewCacheTest` objects on top of the base `CCoinsViewTest`,  and  check also the dynamic memory usage with anchors and nullifiers.

File: src/coins.h [hunk 1]
```diff
@@ -109,6 +110,27 @@ class SaltedOutpointHasher
...
+    /**
+     * This *must* return size_t. With Boost 1.46 on 32-bit systems the
+     * unordered_map will behave unpredictably if the custom hasher returns a
+     * uint64_t, resulting in failures when syncing the chain (#4634).
+     */
+    size_t operator()(const uint256& txid) const
+    {
+        return SipHashUint256(k0, k1, txid);
+    }
+};
```
[the remark this answers is not in the document]
Comment (furszy) [reply src/coins.h:132]: I missed it.
`txmempool.h` depends on `coins.h`. So, it's in the other way around, `SaltedTxidHasher` can be removed for now.

File: src/coins.h [hunk 2]
```diff
@@ -161,7 +213,11 @@ class CCoinsView
 
     //! Do a bulk modification (multiple Coin changes + BestBlock change).
     //! The passed mapCoins can be modified.
-    virtual bool BatchWrite(CCoinsMap& mapCoins, const uint256& hashBlock);
+    virtual bool BatchWrite(CCoinsMap& mapCoins,
+                            const uint256& hashBlock,
+                            const uint256 &hashSaplingAnchor,
+                            CAnchorsSaplingMap &mapSaplingAnchors,
+                            CNullifiersMap &mapSaplingNullifiers);
```
[the remark this answers is not in the document]
Comment (furszy) [reply src/coins.h:220]: I have it set in that way already :), but it doesn't align the styling on cherry-picked commits, i made all of this work in May/June. Will align it manually.

File: src/txdb.h [hunk 3]
```diff
@@ -76,12 +76,26 @@ class CCoinsViewDB : public CCoinsView
...
     bool Upgrade();
     size_t EstimateSize() const override;
+
+    bool BatchWrite(CCoinsMap& mapCoins,
+                    const uint256& hashBlock,
+                    const uint256 &hashSaplingAnchor,
+                    CAnchorsSaplingMap &mapSaplingAnchors,
+                    CNullifiersMap &mapSaplingNullifiers) override;
+
+    // Sapling
```
[the remark this answers is not in the document]
Comment (furszy) [reply src/txdb.h:91]: yeah, done.

**Comment by furszy** [comment]
updated per feedback.

**Review by Fuzzbawls — APPROVED** [review]
utACK 3b4ad0dc901f41a5701826e6bb5477a2faa59d85

**Review by random-zebra — APPROVED** [review]
utACK 3b4ad0dc901f41a5701826e6bb5477a2faa59d85 and merging...

**Pull request merged**